File: test/card_instance_test.dart
import 'package:flutter_test/flutter_test.dart';
import '../lib/models/card_instance.dart';

void main() {
  group('CardInstance Tests', () {
    test('should create unique instance IDs for same card type', () {
      // Create multiple instances of the same card
      final instance1 = CardInstance(
        instanceId: 'instance-1',
        cardId: 'lightning-bolt',
      );
      final instance2 = CardInstance(
        instanceId: 'instance-2',
        cardId: 'lightning-bolt',
      );
      final instance3 = CardInstance(
        instanceId: 'instance-3',
        cardId: 'lightning-bolt',
      );

      // All instances should have the same card ID
      expect(instance1.cardId, equals('lightning-bolt'));
      expect(instance2.cardId, equals('lightning-bolt'));
      expect(instance3.cardId, equals('lightning-bolt'));

      // But different instance IDs
      expect(instance1.instanceId, isNot(equals(instance2.instanceId)));
      expect(instance2.instanceId, isNot(equals(instance3.instanceId)));
      expect(instance1.instanceId, isNot(equals(instance3.instanceId)));
    });

    test('should compare instances by instanceId not cardId', () {
      final instance1 = CardInstance(
        instanceId: 'instance-1',
        cardId: 'lightning-bolt',
      );
      final instance2 = CardInstance(
        instanceId: 'instance-2',
        cardId: 'lightning-bolt',
      );
      final instance1Copy = CardInstance(
        instanceId: 'instance-1',
        cardId: 'lightning-bolt',
      );

      // Different instances of same card should not be equal
      expect(instance1, isNot(equals(instance2)));

      // Same instance ID should be equal
      expect(instance1, equals(instance1Copy));
    });

    test('should serialize and deserialize correctly', () {
      final instance = CardInstance(
        instanceId: 'test-instance-123',
        cardId: 'fireball',
      );

      final json = instance.toJson();
      expect(json['instanceId'], equals('test-instance-123'));
      expect(json['cardId'], equals('fireball'));

      final deserialized = CardInstance.fromJson(json);
      expect(deserialized.instanceId, equals(instance.instanceId));
      expect(deserialized.cardId, equals(instance.cardId));
    });

    test('should handle a hand with duplicate cards', () {
      // Simulate a hand with 3 Lightning Bolts and 2 Fireballs
      final hand = [
        CardInstance(instanceId: 'inst-1', cardId: 'lightning-bolt'),
        CardInstance(instanceId: 'inst-2', cardId: 'fireball'),
        CardInstance(instanceId: 'inst-3', cardId: 'lightning-bolt'),
        CardInstance(instanceId: 'inst-4', cardId: 'fireball'),
        CardInstance(instanceId: 'inst-5', cardId: 'lightning-bolt'),
      ];

      // Count cards by type
      final cardCounts = <String, int>{};
      for (final instance in hand) {
        cardCounts[instance.cardId] = (cardCounts[instance.cardId] ?? 0) + 1;
      }

      expect(cardCounts['lightning-bolt'], equals(3));
      expect(cardCounts['fireball'], equals(2));

      // Each instance should be unique
      final instanceIds = hand.map((e) => e.instanceId).toSet();
      expect(instanceIds.length, equals(5));
    });

    test('should allow selecting specific duplicates for discard', () {
      final hand = [
        CardInstance(instanceId: 'inst-1', cardId: 'lightning-bolt'),
        CardInstance(instanceId: 'inst-2', cardId: 'lightning-bolt'),
        CardInstance(instanceId: 'inst-3', cardId: 'lightning-bolt'),
      ];

      // Select only the second Lightning Bolt for discard
      final selectedForDiscard = <String>{'inst-2'};

      // Check which cards are selected
      final discardedCards = hand
          .where((card) => selectedForDiscard.contains(card.instanceId))
          .toList();

      expect(discardedCards.length, equals(1));
      expect(discardedCards.first.instanceId, equals('inst-2'));

      // The other Lightning Bolts should not be selected
      final remainingCards = hand
          .where((card) => !selectedForDiscard.contains(card.instanceId))
          .toList();

      expect(remainingCards.length, equals(2));
      expect(remainingCards.any((c) => c.instanceId == 'inst-1'), isTrue);
      expect(remainingCards.any((c) => c.instanceId == 'inst-3'), isTrue);
    });
  });
}

File: backend/internal/domain/card_instance_test.go
package domain

import (
	"testing"
)

func TestCardInstance(t *testing.T) {
	t.Run("should create unique instances for same card type", func(t *testing.T) {
		// Create multiple instances of the same card
		instance1 := NewCardInstance("lightning-bolt")
		instance2 := NewCardInstance("lightning-bolt")
		instance3 := NewCardInstance("lightning-bolt")

		// All instances should have the same card ID
		if instance1.CardID != "lightning-bolt" {
			t.Errorf("Expected CardID to be 'lightning-bolt', got %s", instance1.CardID)
		}
		if instance2.CardID != "lightning-bolt" {
			t.Errorf("Expected CardID to be 'lightning-bolt', got %s", instance2.CardID)
		}
		if instance3.CardID != "lightning-bolt" {
			t.Errorf("Expected CardID to be 'lightning-bolt', got %s", instance3.CardID)
		}

		// But different instance IDs
		if instance1.InstanceID == instance2.InstanceID {
			t.Error("Instance IDs should be unique")
		}
		if instance2.InstanceID == instance3.InstanceID {
			t.Error("Instance IDs should be unique")
		}
		if instance1.InstanceID == instance3.InstanceID {
			t.Error("Instance IDs should be unique")
		}
	})

	t.Run("should handle a deck with duplicate cards", func(t *testing.T) {
		// Simulate a draw pile with 3 Lightning Bolts and 2 Fireballs
		drawPile := []CardInstance{
			NewCardInstance("lightning-bolt"),
			NewCardInstance("fireball"),
			NewCardInstance("lightning-bolt"),
			NewCardInstance("fireball"),
			NewCardInstance("lightning-bolt"),
		}

		// Count cards by type
		cardCounts := make(map[CardID]int)
		for _, instance := range drawPile {
			cardCounts[instance.CardID]++
		}

		if cardCounts["lightning-bolt"] != 3 {
			t.Errorf("Expected 3 lightning-bolt cards, got %d", cardCounts["lightning-bolt"])
		}
		if cardCounts["fireball"] != 2 {
			t.Errorf("Expected 2 fireball cards, got %d", cardCounts["fireball"])
		}

		// Each instance should be unique
		instanceIDs := make(map[CardInstanceID]bool)
		for _, instance := range drawPile {
			if instanceIDs[instance.InstanceID] {
				t.Error("Found duplicate instance ID")
			}
			instanceIDs[instance.InstanceID] = true
		}

		if len(instanceIDs) != 5 {
			t.Errorf("Expected 5 unique instances, got %d", len(instanceIDs))
		}
	})

	t.Run("should allow discarding specific duplicates", func(t *testing.T) {
		// Create a hand with 3 Lightning Bolts
		hand := []CardInstance{
			NewCardInstance("lightning-bolt"),
			NewCardInstance("lightning-bolt"),
			NewCardInstance("lightning-bolt"),
		}

		// Select only the second Lightning Bolt for discard
		selectedForDiscard := hand[1].InstanceID

		// Simulate discarding
		var remainingHand []CardInstance
		var discardPile []CardInstance

		for _, card := range hand {
			if card.InstanceID == selectedForDiscard {
				discardPile = append(discardPile, card)
			} else {
				remainingHand = append(remainingHand, card)
			}
		}

		// Check results
		if len(discardPile) != 1 {
			t.Errorf("Expected 1 card in discard pile, got %d", len(discardPile))
		}
		if len(remainingHand) != 2 {
			t.Errorf("Expected 2 cards remaining in hand, got %d", len(remainingHand))
		}

		// All cards should still be Lightning Bolts
		for _, card := range remainingHand {
			if card.CardID != "lightning-bolt" {
				t.Errorf("Expected all remaining cards to be lightning-bolt, got %s", card.CardID)
			}
		}
	})
}